src/storage/CompactionFilter.h

```diff
@@ -8,40 +8,100 @@
-                std::string* new_val, bool* value_changed) const override {
-        UNUSED(level);
-        UNUSED(key);
-        UNUSED(old_val);
-        UNUSED(new_val);
-        UNUSED(value_changed);
-        LOG(ERROR) << "NebulaCompactionFilter not supported yet";
+    explicit NebulaCompactionFilter(meta::SchemaManager* schemaMan)
+        : schemaMan_(schemaMan) {
+        CHECK_NOTNULL(schemaMan_);
+    }
+
+    bool filter(GraphSpaceID spaceId,
+                const folly::StringPiece& key,
+                const folly::StringPiece& val) const override {
+        if (NebulaKeyUtils::isDataKey(key)) {
+            if (!schemaValid(spaceId, key)) {
+                return true;
+            }
+            if (!ttlValid(spaceId, val)) {
+                VLOG(3) << "TTL invalid for key " << key;
+                return true;
+            }
+            if (filterVersions(key)) {
+                VLOG(3) << "Extra versions has been filtered!";
+                return true;
+            }
+        } else if (NebulaKeyUtils::isIndexKey(key)) {
+            if (!indexValid(key)) {
+                VLOG(3) << "Index invalid for the key " << key;
+                return true;
+            }
+        } else {
+            VLOG(3) << "Skip the system key inside, key " << key;
+        }
+        return false;
+    }
+
+    bool schemaValid(GraphSpaceID spaceId, const folly::StringPiece& key) const {
+        if (NebulaKeyUtils::isVertex(key)) {
+            auto tagId = NebulaKeyUtils::getTagId(key);
+            if (schemaMan_->getNewestTagSchemaVer(spaceId, tagId) < 0) {
+                VLOG(3) << "Space " << spaceId << ", Tag " << tagId << " invalid";
+                return false;
+            }
+        } else {
+            CHECK(NebulaKeyUtils::isEdge(key));
+            auto edgeType = NebulaKeyUtils::getEdgeType(key);
+            if (edgeType < 0) {
+                edgeType = -edgeType;
+            }
+            if (schemaMan_->getNewestEdgeSchemaVer(spaceId, edgeType) < 0) {
+                VLOG(3) << "Space " << spaceId << ", EdgeType " << edgeType << " invalid";
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool ttlValid(GraphSpaceID, const folly::StringPiece&) const {
+        return true;
+    }
+
+    bool filterVersions(const folly::StringPiece& key) const {
+        folly::StringPiece keyWithNoVersion = NebulaKeyUtils::keyWithNoVersion(key);
+        if (keyWithNoVersion == lastKeyWithNoVerison_) {
+            // TODO(heng): we could support max-versions configuration in schema if needed.
+            return true;
+        }
+        lastKeyWithNoVerison_ = keyWithNoVersion.str();
         return false;
     }
 
-    const char* Name() const override {
-        return "NebulaCompactionFilter";
+    bool indexValid(const folly::StringPiece&) const {
+        return true;
     }
+
+private:
+    mutable std::string lastKeyWithNoVerison_;
```

Comment:
Is the filter called from a single thread? If not, `lastKeyWithNoVersion` shoud be protected.

Reply:
Yes,  sub-compaction is closed by default.  Of course, we could check the config again in code.